cobc/replace.c

```diff
@@ -0,0 +1,841 @@
+   stream of `replace`.
+
+   Once replace operations have been performed on this second stream,
+   `cb_ppecho_direct()` (in pplex.l) is used to output the final
+   tokens.
+
+   The states of both streams are stored in a struct
+   `cb_replacement_state`, and `add_text_to_replace` calls the
+   function `do_replace()` to perform the replacement on a given
+   stream.
+ */
+
+/* Uncomment the following lines to have a trace of replacements.
+   It uses macros WITH_DEPTH that adds an additional argument to every
+   function to keep the depth of the recursion. */
+
+/* #define DEBUG_REPLACE_TRACE */
+/* #define DEBUG_REPLACE */
+
+#ifdef DEBUG_REPLACE_TRACE
+#define DEBUG_REPLACE
+#endif
+
+struct cb_token_list {
+	struct cb_token_list	*next;			/* next pointer */
+	struct cb_token_list	*last;
+
+	 /* The text in the source to be matched. Most of the time, it
+	  *  directly what appears in the source file, but it may also
+	  *  be a simplified version, typically for spaces, in which
+	  *  case the exact text is stored in the `token` field (to be
+	  *  used if no replacement is performed) */
+	const char		*text;
+
+	/* NULL most of the time, non-NULL only if the `text` was
+	 * replaced by a simplified version, i.e. space to easy
+	 * testing. */
+	const char		*token;
+};
+
+/* types */
+enum cb_ppecho {
+	CB_PPECHO_DIRECT  = 0, /* direct output */
+	CB_PPECHO_REPLACE = 1, /* output to REPLACE */
+};
+
+struct cb_replacement_state {
+
+	/* The list of tokens that are currently being checked for
+	 * replacements. Empty, unless a partial match occurred. */
+	struct cb_token_list *token_queue ;
+
+	/* We don't queue WORD tokens immediately, because
+	 * preprocessing could create larger words. Instead, we buffer
+	 * WORD tokens (and merge them) until another kind of token
+	 * (SPACE,DELIM,etc.) is received. */
+	const char              *text_prequeue ;
+
+	/* Current list of replacements specified in COPY-REPLACING or
+	 * REPLACE */
+	struct cb_replace_list  *replace_list ;
+
+	/* List of replacements after a partial match that still need
+	 * to be tested. */
+	const struct cb_replace_list  *current_list ;
+
+	/* The next pass to which generated tokens should be passed
+	 * (either REPLACE pass or direct output */
+	enum cb_ppecho           ppecho ;
+
+#ifdef DEBUG_REPLACE
+	const char* name ;
+#endif
+};
+
+
+#ifdef DEBUG_REPLACE_TRACE
+
+#define WITH_DEPTH int depth,
+#define INIT_DEPTH 1,
+#define MORE_DEPTH depth+1,
+
+#define MAX_DEPTH 100
+char depth_buffer[MAX_DEPTH+1];
+#define DEPTH depth_buffer + ( MAX_DEPTH-depth )
+
+#else // DEBUG_REPLACE_TRACE
+
+#define WITH_DEPTH
+#define DEPTH
+#define INIT_DEPTH 
+#define MORE_DEPTH
+
+#endif // DEBUG_REPLACE_TRACE
+
+
+#ifdef DEBUG_REPLACE
+
+#define MAX_TEXT_LIST_STRING 10000
+char text_list_string[MAX_TEXT_LIST_STRING];
+
+static
+char * string_of_list(const struct cb_text_list *list)
+{
+	int pos = 1;
+	text_list_string[0] = '[';
+
+	for(; list != NULL; list = list->next){
```

Comment:
please add a comment to this code and/or function doc, from looking at it I have no clue what this may does

Reply:
It prints a list of tokens/texts into a simple string, using always the same statically allocated string... but only when the DEBUG macro is set for debugging.